service/history/history_engine2_test.go

```diff
@@ -1354,104 +1257,228 @@ func (s *engine2Suite) TestStartWorkflowExecution_NotRunning_PrevSuccess() {
+		})
+
+		s.Run("with success", func() {
+			s.Run("and id reuse policy is ALLOW_DUPLICATE", func() {
+				s.mockExecutionMgr.EXPECT().CreateWorkflowExecution(gomock.Any(), brandNewExecutionRequest).
+					Return(nil, makeCurrentWorkflowConditionFailedError(prevRequestID))
+				s.mockExecutionMgr.EXPECT().CreateWorkflowExecution(gomock.Any(), updateExecutionRequest).
+					Return(tests.CreateWorkflowExecutionResponse, nil)
+
+				resp, err := s.historyEngine.StartWorkflowExecution(
+					metrics.AddMetricsContext(context.Background()),
+					makeStartRequest(enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE))
+
+				s.NoError(err)
+				s.NotEqual(prevRunID, resp.GetRunId())
 			})
 
-			if expecedErrs[j] {
-				if _, ok := err.(*serviceerror.WorkflowExecutionAlreadyStarted); !ok {
-					s.Fail("return err is not *serviceerror.WorkflowExecutionAlreadyStarted")
-				}
+			s.Run("and id reuse policy ALLOW_DUPLICATE_FAILED_ONLY", func() {
+				s.mockExecutionMgr.EXPECT().CreateWorkflowExecution(gomock.Any(), brandNewExecutionRequest).
+					Return(nil, makeCurrentWorkflowConditionFailedError(prevRequestID))
+
+				resp, err := s.historyEngine.StartWorkflowExecution(
+					metrics.AddMetricsContext(context.Background()),
+					makeStartRequest(enumspb.WORKFLOW_ID_REUSE_POLICY_ALLOW_DUPLICATE_FAILED_ONLY))
+
+				var expectedErr *serviceerror.WorkflowExecutionAlreadyStarted
+				s.ErrorAs(err, &expectedErr)
 				s.Nil(resp)
-			} else {
-				s.Nil(err)
-				s.NotNil(resp)
+			})
+
+			s.Run("and id reuse policy REJECT_DUPLICATE", func() {
+				s.mockExecutionMgr.EXPECT().CreateWorkflowExecution(gomock.Any(), brandNewExecutionRequest).
+					Return(nil, makeCurrentWorkflowConditionFailedError(prevRequestID))
+
+				resp, err := s.historyEngine.StartWorkflowExecution(
+					metrics.AddMetricsContext(context.Background()),
+					makeStartRequest(enumspb.WORKFLOW_ID_REUSE_POLICY_REJECT_DUPLICATE))
+
+				var expectedErr *serviceerror.WorkflowExecutionAlreadyStarted
+				s.ErrorAs(err, &expectedErr)
+				s.Nil(resp)
+			})
+		})
+
+		s.Run("with failure", func() {
+			failureStatuses := []enumspb.WorkflowExecutionStatus{
+				enumspb.WORKFLOW_EXECUTION_STATUS_FAILED,
+				enumspb.WORKFLOW_EXECUTION_STATUS_CANCELED,
+				enumspb.WORKFLOW_EXECUTION_STATUS_TERMINATED,
+				enumspb.WORKFLOW_EXECUTION_STATUS_TIMED_OUT,
 			}
-		}
-	}
+
+			for _, status := range failureStatuses {
+				makeCurrentWorkflowConditionFailedError := func(
+					requestID string,
+				) *persistence.CurrentWorkflowConditionFailedError {
+					return &persistence.CurrentWorkflowConditionFailedError{
+						Msg:              "random message",
+						RequestID:        requestID,
+						RunID:            prevRunID,
+						State:            enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED,
+						Status:           status,
+						LastWriteVersion: lastWriteVersion,
+					}
+				}
+
+				s.Run(fmt.Sprintf("status %v", status), func() {
+					s.Run("and id reuse policy ALLOW_DUPLICATE", func() {
```

Comment:
:+1:. You should add a screenshot of test runner with all this nested tests are concatenated to plain English sentence.

Reply:
Goland only shows them hierarchical:

<img width="524" alt="Screenshot 2024-03-05 at 10 15 42 AM" src="https://github.com/temporalio/temporal/assets/159852/9e7a9833-4472-4b65-930a-c4a23942cf64">